src/validations/html/custom.js

```diff
@@ -0,0 +1,144 @@
+import {List, Map, Record, Set} from 'immutable';
+import {SAXParser} from 'parse5';
+import voidElements from 'void-elements';
+import Validator from '../Validator';
+
+// An immutable.Record wrapper of parse5.MarkupData.Location, whose line and
+// col are 1-indexed.
+const Location = Record({line: 1, col: 1});  // eslint-disable-line new-cap
+
+function findMismatchedCloseTags(parser) {
```

Comment:
At a high level, I’d propose restructuring the way you store the state during the tree walk. In particular I would hold on to three collections, each of which corresponds to one of the errors this module can generate:

1. Map of opening tags that don’t have a properly-placed closing tag, and for which we haven’t yet seen a misplaced closing tag, keyed by tag name, with a queue of locations (`unclosed-tag-by-mismatched-close-tag`)
2. Set of pairs of opening/closing tags where the closing tag is in the wrong place (`misplaced-close-tag`)
3. Set of closing tags that don’t match any opening tag (`unexpected-close-tag`)

Any time we encounter a mismatch:

* Add all unclosed tags to collection 1
* If the closing tag we have encountered does not match an opening tag in the stack (`!open`):
  * If the closing tag we have encountered matches an entry in collection 1, remove that entry from collection 1, and add the open/close pair to collection 2
  * Otherwise, add the closing tag to collection 3

Obviously this structure has a lot in common with what you’ve got now, but I think it’ll be easier to reason about. In particular, once the initial tree walk is finished, we have a final accounting of the three error types we can generate—there is no need for a second pass to turn raw observations into specific errors. As a corollary, understanding the entire process requires understanding only the types of errors the module emits; there is no need to additionally understand the structure and intent of a set of intermediate data structures (something I found fairly challenging when taking this code in just now).

Definitely possible I am missing some nuance here that my proposal doesn’t account for, let me know what you think!